src/os_darwin/os_futex.c

```diff
@@ -0,0 +1,83 @@
+/*-
+ * Copyright (c) 2024-present MongoDB, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#include "wt_internal.h"
+#include <sys/errno.h>
+
+#define PRIVATE 1
+#include <ulock.h>
+#undef PRIVATE
+
+/*
+ * __wt_futex_wait --
+ *     Wait on the futex. The timeout is in microseconds and MUST be greater than zero.
+ */
+int
+__wt_futex_wait(
+  volatile WT_FUTEX_WORD *addr, WT_FUTEX_WORD expected, time_t usec, WT_FUTEX_WORD *wake_valp)
+{
+    WT_DECL_RET;
+    uint64_t nsec;
+
+    /* Check for overflow? */
+    nsec = (uint64_t)usec * 1000;
+
+    ret = __ulock_wait2(UL_COMPARE_AND_WAIT_SHARED | ULF_NO_ERRNO, (void *)addr, expected, nsec, 0);
+    if (ret >= 0) {
+        *wake_valp = __atomic_load_n(addr, __ATOMIC_SEQ_CST);
+        ret = 0;
+    } else if (ret == -EFAULT) {
+        /*
+         * Page containing futex was paged out. The value is unknown so assume it has not changed,
+         * and emulate a spurious wakeup to prompt the caller to retry.
+         */
+        *wake_valp = expected;
```

Comment:
I'm not sure how this "assume unchanged" interacts with other parts of the system. Is it sane for a caller to rely on `expected` being correct even in the event of a spurious timeout? Not sure I'm articulating this well, perhaps an IRL conversation would be better.

Reply:
comment rewritten